src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs

```diff
@@ -0,0 +1,441 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed partial class NullableWalker
+    {
+        /// <summary>
+        /// Learn something about the input from a test of a given expression against a given pattern.  The given
+        /// state is updated to note that any slots that are tested against `null` may be null.
+        /// </summary>
+        /// <returns>true if there is a top-level explicit null check</returns>
+        private void LearnFromAnyNullPatterns(
+            BoundExpression expression,
+            BoundPattern pattern)
+        {
+            int slot = MakeSlot(expression);
+            LearnFromAnyNullPatterns(slot, expression.Type, pattern);
+        }
+
+        /// <summary>
+        /// Learn from any constant null patterns appearing in the pattern.
+        /// </summary>
+        /// <param name="inputType">Tye type of the input expression (before nullable analysis).
+        /// Used to determine which types can contain null.</param>
+        /// <returns>true if there is a top-level explicit null check</returns>
+        private void LearnFromAnyNullPatterns(
+            int inputSlot,
+            TypeSymbol inputType,
+            BoundPattern pattern)
+        {
+            if (inputSlot <= 0)
+                return;
+
+            switch (pattern)
+            {
+                case BoundConstantPattern cp:
+                    bool isExplicitNullCheck = cp.Value.ConstantValue == ConstantValue.Null;
+                    if (isExplicitNullCheck)
+                    {
+                        LearnFromNullTest(inputSlot, inputType, ref this.State);
+                    }
+                    break;
+                case BoundDeclarationPattern _:
+                case BoundDiscardPattern _:
+                case BoundITuplePattern _:
+                    break; // nothing to learn
+                case BoundRecursivePattern rp:
+                    {
+                        // for positional part: we only learn from tuples (not Deconstruct)
```

Comment:
>Deconstruct [](start = 79, length = 11)

Why not?

Reply:
Because we track fields and properties, not the values generated by calling a method with out parameters.